keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,236 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+Reclaim policy associated with the Persistent Volume is currently ignored under certain circumstance. For a `Bound`
+PV-PVC pair the ordering of PV-PVC deletion determines whether the PV delete reclaim policy is honored. The PV honors
+the reclaim policy if the PVC is deleted prior to deleting the PV, however, if the PV is deleted prior to deleting the
+PVC then the Reclaim policy is not exercised. As a result of this behavior, the associated storage asset in the
+external infrastructure is not removed.
+
+## Motivation
+Prevent volumes from being leaked by honoring the PV Reclaim policy after the `Bound` PVC is also deleted.
+
+## Proposal
+To better understand the existing issue we will initially walk through existing behavior when a Bound PV is deleted prior
+to deleting the PVC.
+
+
+```
+kubectl delete pv <pv-name>
+```
+On deleting a `Bound` PV, a `deletionTimestamp` is added on to the PV object, this triggers an update which is consumed
+by the PV-PVC-controller. The PV-PVC-controller observes that the PV is still associated with a PVC, and the PVC hasn't
+been deleted, as a result of this, the PV-PVC-controller attempts to update the PV phase to `Bound` state. Assuming that
+the PV was already bound previously, this update eventually amounts to a no-op.
+
+The PVC is deleted at a later point of time.
+```
+kubectl -n <namespace> delete pvc <pvc-name>
+```
+
+1. A `deletionTimestamp` is added on the PVC object.
+2. The PVC-protection-controller picks the update, verifies if `deletionTimestamp` is present and there are no pods that
+are currently using the PVC.
+3. If there are no Pods using the PVC, the PVC finalizers are removed, eventually triggering a PVC delete event.
+4. The PV-PVC-controller processes delete PVC event, leading to removal of the PVC from it's in-memory cache followed
+by triggering an explicit sync on the PV.
+5. The PV-PVC-controller processes the triggered explicit sync, here, it observes that the PVC is not longer available
+and updates the PV phase to `Released` state.
+6. If the PVC-protection-controller picks the update, it observes that there is a `deletionTimestamp` and the PV is not
+in a `Bound` phase, this causes the finalizers to be removed.
+7. This is followed by PV-PVC-controller initiating the reclaim volume workflows.
+8. The reclaim volume workflow observes the `persistentVolumeReclaimPolicy` as `Delete` and schedules a volume deletion.
+9. During delete volume operation, it is observed that there is already a `deletionTimestamp`, and the call is never
+forwarded to the CSI driver to delete the underlying volume.
+   
+As observed from the description above, (9) prevents the volume deletion to be forwarded to the plugin.
+
+Preventing the check should allow the volume deletion.
+
+
+### Risks and Mitigations
+N/A
```

Comment:
This is not N/A. At the very minimum, there's a risk of accidental volume completion if we change the default behaviour and people are relying on the current behaviour.

Reply:
Done